File: 102ShiftedBinarySearch/shifted_binary_search.rb
# Solution 1 - recursion
# O(log(n)) time / O(log(n)) space

def shifted_binary_search(arr, target)
  shifted_binary_search_helper(arr, target, 0, arr.length - 1)
end

def shifted_binary_search_helper(arr, target, left, right)
  return -1 if left > right

  middle = ((left + right) / 2).floor
  potential_match = arr[middle]
  left_num = arr[left]
  right_num = arr[right]

  if target == potential_match
    middle
  elsif left_num <= potential_match
    if target < potential_match && target >= left_num
      shifted_binary_search_helper(arr, target, left, middle - 1)
    else
      shifted_binary_search_helper(arr, target, middle + 1, right)
    end
  elsif target > potential_match && target <= right_num
    shifted_binary_search_helper(arr, target, middle + 1, right)
  else
    shifted_binary_search_helper(arr, target, left, middle - 1)
  end
end

# Solution 2 - iteration
# O(log(n)) time / O(1) space

def shifted_binary_search(arr, target)
  shifted_binary_search_helper(arr, target, 0, arr.length - 1)
end

def shifted_binary_search_helper(arr, target, left, right)
  while left <= right
    middle = ((left + right) / 2).floor
    potential_match = arr[middle]
    left_num = arr[left]
    right_num = arr[right]

    if target == potential_match
      middle
    elsif left_num <= potential_match
      if target < potential_match && target >= left_num
        right = middle - 1
      else
        left = middle + 1
      end
    elsif target > potential_match && target <= right_num
      left = middle + 1
    else
      right = middle - 1
    end
  end
  -1
end

File: 102ShiftedBinarySearch/shiftedBinarySearch.js
// Solution 1 - recursion
// O(log(n)) time / O(log(n)) space

const shiftedBinarySearch = (array, target) => {
  return shiftedBinarySearchHelper(array, target, 0, array.length - 1);
};

const shiftedBinarySearchHelper = (array, target, left, right) => {
  if (left > right) return -1;

  const middle = Math.floor((left + right) / 2);
  const potentialMatch = array[middle];
  const leftNum = array[left];
  const rightNum = array[right];

  if (target === potentialMatch) {
    return middle;
  } else if (leftNum <= potentialMatch) {
    if (target < potentialMatch && target >= leftNum) {
      return shiftedBinarySearchHelper(array, target, left, middle - 1);
    } else {
      return shiftedBinarySearchHelper(array, target, middle + 1, right);
    }
  } else {
    if (target > potentialMatch && target <= rightNum) {
      return shiftedBinarySearchHelper(array, target, middle + 1, right);
    } else {
      return shiftedBinarySearchHelper(array, target, left, middle - 1);
    }
  }
};

// Solution 2 - iteration
// O(log(n)) time / O(1) space

const shiftedBinarySearch = (array, target) => {
  return shiftedBinarySearchHelper(array, target, 0, array.length - 1);
};

const shiftedBinarySearchHelper = (array, target, left, right) => {
  while (left <= right) {
    const middle = Math.floor((left + right) / 2);
    const potentialMatch = array[middle];
    const leftNum = array[left];
    const rightNum = array[right];

    if (target === potentialMatch) {
      return middle;
    } else if (leftNum <= potentialMatch) {
      if (target < potentialMatch && target >= leftNum) {
        right = middle - 1;
      } else {
        left = middle + 1;
      }
    } else {
      if (target > potentialMatch && target <= rightNum) {
        left = middle + 1;
      } else {
        right = middle - 1;
      }
    }
  }
  return -1;
};
